components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/ApplicationAuthenticatorService.java

```diff
@@ -121,4 +134,73 @@ public void removeRequestPathAuthenticator(RequestPathAuthenticatorConfig authen
             requestPathAuthenticators.remove(authenticator);
         }
     }
+
+    public LocalAuthenticatorConfig createUserDefinedLocalAuthenticator(LocalAuthenticatorConfig authenticatorConfig,
+              AuthenticatorType type, String tenantDomain) throws AuthenticatorMgtException {
+
+        LocalAuthenticatorConfig config = getLocalAuthenticatorByName(authenticatorConfig.getName());
+        if (config != null) {
+            ErrorMessages error = ErrorMessages.ERROR_AUTHENTICATOR_ALREADY_EXIST;
+            throw new AuthenticatorMgtClientException(error.getCode(), error.getMessage(),
+                    String.format(error.getDescription(), authenticatorConfig.getName()));
+        }
+        validateAuthenticatorDefinedByType(authenticatorConfig);
+
+        LocalAuthenticatorConfig createdConfig = CACHE_BACKED_DAO.addUserDefinedLocalAuthenticator(
+                authenticatorConfig, IdentityTenantUtil.getTenantId(tenantDomain), type);
+
+        localAuthenticators.add(createdConfig);
+        return createdConfig;
+    }
+
+    public LocalAuthenticatorConfig updateUserDefinedLocalAuthenticator(LocalAuthenticatorConfig authenticatorConfig,
+            String tenantDomain) throws AuthenticatorMgtException {
+
+        LocalAuthenticatorConfig existingAuthenticatorConfig = getExistingAuthenticatorConfig(
+                authenticatorConfig.getName());
```

Comment:
If this is initiating a db call, it's not a good practice to retrieve and update IMO

Reply:
We need to retrieve the authenticator to check whether there is an exitsing authenticator and check its definedBy type is `USER`.